File: backends/spring-java/src/main/java/dev/stackverse/backend/moderation/ReportRepository.java
package dev.stackverse.backend.moderation;

import jakarta.persistence.LockModeType;
import java.util.List;
import java.util.UUID;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Lock;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

public interface ReportRepository extends JpaRepository<Report, UUID> {
    @Lock(LockModeType.PESSIMISTIC_WRITE)
    Report findForUpdateById(UUID id);

    @Query("select r.bookmarkId from Report r where r.id = :id")
    UUID findBookmarkIdById(@Param("id") UUID id);

    @Lock(LockModeType.PESSIMISTIC_WRITE)
    List<Report> findForUpdateByBookmarkIdAndStatusOrderByIdAsc(UUID bookmarkId, ReportStatus status);

    boolean existsByBookmarkIdAndReporterAndStatus(UUID bookmarkId, String reporter, ReportStatus status);

    boolean existsByBookmarkIdAndReporterAndStatusAndIdNot(UUID bookmarkId, String reporter, ReportStatus status, UUID id);

    Page<Report> findByStatus(ReportStatus status, Pageable pageable);

    Page<Report> findByReporter(String reporter, Pageable pageable);

    Page<Report> findByReporterAndStatus(String reporter, ReportStatus status, Pageable pageable);

    long countByStatus(ReportStatus status);
}

File: backends/spring-kotlin/src/main/kotlin/dev/stackverse/backend/moderation/ReportRepository.kt
package dev.stackverse.backend.moderation

import jakarta.persistence.LockModeType
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Lock
import java.util.UUID

interface ReportRepository : JpaRepository<Report, UUID> {

    /**
     * Row-locked read for the mutating paths: reporter edit/withdraw and
     * moderator resolution all check the status before acting, and without a
     * lock the loser of a race flushes its stale snapshot over the winner's
     * resolution (or deletes a just-resolved report).
     */
    @Lock(LockModeType.PESSIMISTIC_WRITE)
    fun findForUpdateById(id: UUID): Report?

    fun existsByBookmarkIdAndReporterAndStatus(bookmarkId: UUID, reporter: String, status: ReportStatus): Boolean

    fun findByStatus(status: ReportStatus, pageable: Pageable): Page<Report>

    fun findByReporter(reporter: String, pageable: Pageable): Page<Report>

    fun findByReporterAndStatus(reporter: String, status: ReportStatus, pageable: Pageable): Page<Report>

    fun findByBookmarkIdAndStatus(bookmarkId: UUID, status: ReportStatus): List<Report>

    fun countByStatus(status: ReportStatus): Long
}
